### DISCUSSION TITLE
How does Wasm-opt's 0s/0z work?

### DISCUSSION THREAD

**Discussion by Simarkohli24** [WebAssembly/binaryen#5014] [Q&A]
Looking into optimizing my WebAssembly code. Noticed a size-focused parameter, so wanted to know more about what exactly it's doing. 

What are the optimization transforms that Wasm-opt will do?

**Comment by tlively** [comment]
Binaryen's optimizations are a mix of classic compiler optimizations like dead code elimination, common subexpression elimination, inlining, etc. and WebAssembly-specific optimizations like memory packing, removing unused branch instructions and control flow labels, and reordering things to minimize the code size of indices. Using `-Os` or `-Oz` largely runs the same optimization passes as other optimization modes, but tells those passes only to do transformations that would make code size smaller.

You can see the full list of available passes [here](https://github.com/WebAssembly/binaryen/tree/main/src/passes) or with `wasm-opt --help` and you can see the specific sequence of passes wasm-opt runs [here](https://github.com/WebAssembly/binaryen/blob/31f1ba707a27f3000cecba87926980238f82414a/src/passes/pass.cpp#L458-L462).
